tests/main/tests/integration/record-array-test.js

```diff
@@ -400,4 +400,100 @@ module('unit/record-array - RecordArray', function (hooks) {
 
     assert.true(people.isLoaded, 'The array is now loaded');
   });
+
+  test('a hasMany record array should only remove object(s) if found in collection', async function (assert) {
+    class AsyncTag extends Model {
+      @hasMany('person', { async: true, inverse: 'tag' })
+      people;
+    }
+
+    store.push({
+      data: [
+        {
+          type: 'person',
+          id: '1',
+          attributes: {
+            name: 'Scumbag Dale',
+          },
+        },
+        {
+          type: 'person',
+          id: '2',
+          attributes: {
+            name: 'Scumbag Tom',
+          },
+        },
+        {
+          type: 'person',
+          id: '3',
+          attributes: {
+            name: 'Scumbag Ross',
+          },
+        },
+        {
+          type: 'tag',
+          id: '1',
+          relationships: {
+            people: {
+              data: [
+                { type: 'person', id: '1' },
+                { type: 'person', id: '2' },
+              ],
+            },
+          },
+        },
+      ],
+    });
+
+    let tag = await store.findRecord('tag', '1');
+    let recordArray = tag.people;
+    let scumbagInRecordArray = await store.findRecord('person', '1');
+    let scumbagNotInRecordArray = await store.findRecord('person', '3');
+
+    recordArray.removeObject(scumbagNotInRecordArray);
+
+    assert.strictEqual(
+      recordArray.length,
+      2,
+      'sync record array unchanged after attempting to remove object not found in collection'
+    );
+
+    recordArray.removeObject(scumbagInRecordArray);
+
+    let didRemoveObject = recordArray.length === 1 && !recordArray.includes(scumbagInRecordArray);
+    assert.true(didRemoveObject, 'sync record array successfully removed expected object from collection');
+
+    recordArray.push(scumbagInRecordArray);
+
+    let scumbagsToRemove = [scumbagInRecordArray, scumbagNotInRecordArray];
+    recordArray.removeObjects(scumbagsToRemove);
+
+    didRemoveObject = recordArray.length === 1 && !recordArray.includes(scumbagInRecordArray);
+    assert.true(didRemoveObject, 'sync record array only removes objects in list that are found in collection');
+
+    recordArray.push(scumbagInRecordArray);
+    this.owner.unregister('model:tag');
+    this.owner.register('model:tag', AsyncTag);
```

Comment:
we should do this as a separate test, unregister/register with a booted application isn't safe